Do not copy bytes for cached serializations

linxiulei opened the pull request:
Since cachingObject has the encoded data cached and they are not supposed to change. It's memory efficient to just copy the slice references.


#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This PR reduces memory overhead for watch connections

#### Which issue(s) this PR fixes:
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

linxiulei committed:
Do not copy bytes for cached serializations

Since cachingObject has the encoded data cached and they are not
supposed to change. It's memory efficient to just copy the slice
references.

Signed-off-by: Eric Lin <exlin@google.com>

linxiulei commented:
/cc @wojtek-t

aojea commented:
do we have some benchmarks that proves the claim and show the difference?

https://pkg.go.dev/golang.org/x/perf/cmd/benchstat

linxiulei commented:
> do we have some benchmarks that proves the claim and show the difference?
> 
> https://pkg.go.dev/golang.org/x/perf/cmd/benchstat

I don't know how to incorporate this tool with the changes inside kubernetes but I did manual tests with apiserver and noticed the differences.

If we want to just compare the difference between `SpliceBuffer` and `bytes.Buffer`, the difference should be evident as the former just copies the slice reference and latter copies every bytes

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/caching_object.go:
```diff
@@ -136,6 +137,7 @@ func (o *cachingObject) getSerializationResult(id runtime.Identifier) *serializa
 func (o *cachingObject) CacheEncode(id runtime.Identifier, encode func(runtime.Object, io.Writer) error, w io.Writer) error {
 	result := o.getSerializationResult(id)
 	result.once.Do(func() {
+
```
nit: remove this empty line

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/caching_object.go:
```diff
@@ -148,7 +150,12 @@ func (o *cachingObject) CacheEncode(id runtime.Identifier, encode func(runtime.O
 	}
-	_, err := w.Write(result.raw)
+	var err error
+	if b, support := w.(util.Splice); support {
```
nit:
```
if b, ok := w.(util.Splice); support {
  b.Splice(result.raw)
  return nil
}
_, err := w.Write(result.raw)
return err
```

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/util/buf.go:
```diff
@@ -0,0 +1,63 @@
+limitations under the License.
+*/
+
+package util
```
I'm not a huge fan of `util` packages.

I would put it next to allocator in https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/apimachinery/pkg/runtime

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/util/buf.go:
```diff
@@ -0,0 +1,63 @@
+// efficiency purpose. Therefore, the caller should make sure the underlying
+// data is not changed later.
+type Splice interface {
+	Splice([]byte)
```
let's make it also implement the io.Writer interface

wojtek-t replied:
I guess this also needs to implement Reset

wojtek-t replied:
I guess this also needs to implement Reset

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/util/buf.go:
```diff
@@ -0,0 +1,63 @@
+}
+
+// A SpliceBuffer implements Splice and io.Writer interfaces.
+type SpliceBuffer struct {
```
Let's make it a private thing - it's an implementation details.

And let's only expose a `NewSpliceBuffer() Splice` function

jiahuif commented:
/assign @wojtek-t
Would you kindly follow up with this PR. Thank you.
/triage accepted

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/buf.go:
```diff
@@ -0,0 +1,71 @@
+/*
```
nit: let's rename the file to "splice.go"

And please create some unit tests for this file.

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/buf.go:
```diff
@@ -0,0 +1,71 @@
+
+// Bytes returns the data held by the buffer.
+func (sb *spliceBuffer) Bytes() []byte {
+	if sb.buf == nil {
```
nit:
```
if sb.buf != nil {
  return sb.buf.Bytes()
}
return sb.raw
```

seems more intuitive

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/buf.go:
```diff
@@ -0,0 +1,71 @@
+// Reset resets the buffer to be empty.
+func (sb *spliceBuffer) Reset() {
+	sb.raw = make([]byte, 0)
+	sb.buf = nil
```
This doesn't seem correct. I think we should switch to:
```
sb.raw = nil
if sb.buf != nil {
  sb.buf.Reset()
}
```

1. there is no need to create (even with 0 capacity) a slice here
2. for buf, we should really follow the semantics of Reset and reset the buffer

linxiulei replied:
Actually, we do. At least, we should return empty slice in this Bytes(), otherwise json encodes it as null instead of ""

https://go.dev/play/p/assMCnS9-eS

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/splice_test.go:
```diff
@@ -0,0 +1,104 @@
+		testFunc func(sb runtime.Splice, buf *bytes.Buffer)
+	}{
+		{
+			"Basic Write",
```
nit: please use explicit field names, i.e.:
```
name: "Basic Write",
test: func(....) ...
```

it's useful for future extensions if someone wants to add a new field but set it only for subset of usecases

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/splice_test.go:
```diff
@@ -0,0 +1,104 @@
+	testBytes := []byte{0x01, 0x02, 0x03, 0x04}
+
+	testCases := []struct {
+		testname string
```
nit: just call it "name"

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/splice_test.go:
```diff
@@ -0,0 +1,104 @@
+
+	testCases := []struct {
+		testname string
+		testFunc func(sb runtime.Splice, buf *bytes.Buffer)
```
nit: call it `run` maybe

wojtek-t reviewed (commented):
Thanks - just few minor nits in test and then LGTM

wojtek-t commented:
/lgtm
/approve

Thanks!

wojtek-t commented:
The e2e tests are still failing.
This is probably the most evident and clearly related:
```
Kubernetes e2e suite: [It] [sig-api-machinery] Watchers should be able to restart watching from the last resource version observed by the previous watch [Conformance] expand_less | 0s
-- | --
{ failed [FAILED] Watch closed unexpectedly In [It] at: test/e2e/apimachinery/watch.go:433 @ 06/05/23 14:38:30.346 }
```

linxiulei commented:
I think I fixed the issue, which turned out the `SpliceBuffer` was invoked with `Write`, `Reset`, and then  `Splice`, and eventually returned empty data with "Bytes()" because the buf wasn't nil. I also added test to detect that.

This also intrigued me because it meant the `SpliceBuffer` was used by non-cachableObject (using Write) and cacheableObject (using Splice) in the same Watch connection. It worried me at first because the memory used by Write() will persist until the `SpliceBuffer` is freed completely. Fortunately, I found the non-cacheableObject was the BOOKMARK event, which is supposed to not be made cacheable and BOOKMARK objects are very small (64 bytes or less). So it should be fine.

aojea commented:
is not simpler to use a sync.Pool? 

https://pkg.go.dev/sync#example-Pool

wojtek-t commented:
> is not simpler to use a sync.Pool?

What exactly you would like to pool?
The problem is that we generally don't even want to allocate memory and we want to just store the pointer to serialized data. Pooling that won't really help that.
The buffer is only used (as described by Eric above) in a single place for bookmark events, so I think pooling doesn't really buy us anything.

@aojea - I'm lgtm-ing but holding to wait for your comment if you disagree.

/lgtm
/hold

aojea commented:
> What exactly you would like to pool?

/hold cancel

my apologies, I totally got this wrong, I did it and I understood it

```diff
diff --git a/staging/src/k8s.io/apimachinery/pkg/runtime/splice_test.go b/staging/src/k8s.io/apimachinery/pkg/runtime/splice_test.go
index 9d8ca5102dd..33ee0b3159c 100644
--- a/staging/src/k8s.io/apimachinery/pkg/runtime/splice_test.go
+++ b/staging/src/k8s.io/apimachinery/pkg/runtime/splice_test.go
@@ -18,6 +18,7 @@ package runtime_test
 
 import (
        "bytes"
+       "math/rand"
        "testing"
 
        "k8s.io/apimachinery/pkg/runtime"
@@ -119,3 +120,29 @@ func TestSpliceBuffer(t *testing.T) {
        }
 
 }
+
+func BenchmarkWriteSplice(b *testing.B) {
+       data := make([]byte, 1024)
+       rand.Read(data)
+       for n := 0; n < b.N; n++ {
+               sb := runtime.NewSpliceBuffer()
+               sb.Splice(data)
+               if !bytes.Equal(sb.Bytes(), data) {
+                       b.Errorf("unexpected data")
+               }
+               sb.Reset()
+       }
+}
+
+func BenchmarkWriteBuffer(b *testing.B) {
+       data := make([]byte, 1024)
+       rand.Read(data)
+       for n := 0; n < b.N; n++ {
+               buf := &bytes.Buffer{}
+               buf.Write(data)
+               if !bytes.Equal(buf.Bytes(), data) {
+                       b.Errorf("unexpected data")
+               }
+               buf.Reset()
+       }
+}
```

the results are clear now
```
go test -benchmem -run=^$ -bench ^BenchmarkWrite k8s.io/apimachinery/pkg/runtime -v -count 1
goos: linux
goarch: amd64
pkg: k8s.io/apimachinery/pkg/runtime
cpu: AMD EPYC 7B12
BenchmarkWriteSplice
BenchmarkWriteSplice-48         151164015                7.929 ns/op           0 B/op          0 allocs/op
BenchmarkWriteBuffer
BenchmarkWriteBuffer-48          3476392               357.8 ns/op          1024 B/op          1 allocs/op
PASS
ok      k8s.io/apimachinery/pkg/runtime 3.619s
```
